test/e2e/simple_test.go

```diff
@@ -605,3 +605,47 @@ func TestSimpleBuildWithHybridSources(t *testing.T) {
 		t.Fatalf("Error watching build: %v", err)
 	}
 }
+
+// TestCancelledBuild tests that a build with a long running step
+// so that we have time to cancel it.
+func TestCancelledBuild(t *testing.T) {
+	logger := logging.GetContextLogger("TestCancelledBuild")
+
+	clients := buildClients(logger)
+	buildName := "cancelled-build"
+
+	test.CleanupOnInterrupt(func() { teardownBuild(clients, logger, buildName) }, logger)
+	defer teardownBuild(clients, logger, buildName)
+
+	if _, err := clients.buildClient.builds.Create(&v1alpha1.Build{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: buildTestNamespace,
+			Name:      buildName,
+		},
+		Spec: v1alpha1.BuildSpec{
+			Steps: []corev1.Container{{
+				Image: "busybox",
+				Args:  []string{"sleep", "5000"}, // fails.
+			}},
+		},
+	}); err != nil {
+		t.Fatalf("Error creating build: %v", err)
+	}
+
+	// Wait for a little while
+	// FIXME(vdemeester) I would prefer something less flaky-prone
+	time.Sleep(20 * time.Second)
+
+	b, err := clients.buildClient.builds.Get(buildName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error getting build %s: %v", buildName, err)
+	}
+	b.Spec.Status = v1alpha1.BuildSpecStatusCancelled
+	if _, err := clients.buildClient.builds.Update(b); err != nil {
+		t.Fatalf("Error update build status for %s: %v", buildName, err)
+	}
+
+	if _, err := clients.buildClient.watchBuild(buildName); err == nil {
+		t.Fatalf("watchBuild did not return expected `cancelled` error")
```

Comment:
`watchBuild` function returns `build` object so you could consider verifying that build status is updated with expected reason("BuildCancelled") here.